jsonschema pinned version is incompatible, yields `ImportError: cannot import name 'Draft202012Validator'`

### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

The use of jsonschema import by prefect requires jsonschema > 4.0.0, but prefect allows install of jsonschema > 3.2.0

The broken import happens in `prefect/utilities/schema_tools/__init__.py` but the stack is a bit complex and I haven't figured out the exact usage of prefect that causes this import to happen.

### Reproduction

```python3
$ pip install "jsonschema==3.2"
Collecting jsonschema==3.2
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl.metadata (7.8 kB)
Requirement already satisfied: attrs>=17.4.0 in /home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages (from jsonschema==3.2) (22.2.0)Requirement already satisfied: pyrsistent>=0.14.0 in /home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages (from jsonschema==3.2) (0.19.3)
Requirement already satisfied: setuptools in /home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages (from jsonschema==3.2) (65.5.0)
Requirement already satisfied: six>=1.11.0 in /home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages (from jsonschema==3.2) (1.16.0)
Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Installing collected packages: jsonschema
  Attempting uninstall: jsonschema
    Found existing installation: jsonschema 4.21.1
    Uninstalling jsonschema-4.21.1:
      Successfully uninstalled jsonschema-4.21.1
Successfully installed jsonschema-3.2.0

$ ipython

In [1]: from jsonschema.validators import Draft202012Validator
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[1], line 1
----> 1 from jsonschema.validators import Draft202012Validator

ImportError: cannot import name 'Draft202012Validator' from 'jsonschema.validators' (/home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/jsonschema/validators.py)
```


### Error

```python3
def test_slack_errors():
        with pytest.raises(ValueError):
>           my_test_flow_raises_error()

tests/utilities/test_slack.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/flows.py:1224: in __call__
    return enter_flow_run_engine_from_flow_call(
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/engine.py:297: in enter_flow_run_engine_from_flow_call
    retval = from_sync.wait_for_call_in_loop_thread(
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py:243: in wait_for_call_in_loop_thread
    return call.result()
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py:318: in result
    return self.future.result(timeout=timeout)
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py:179: in result
    return self.__get_result()
../../../.pyenv/versions/3.10.9/lib/python3.10/concurrent/futures/_base.py:403: in __get_result
    raise self._exception
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py:389: in _run_async
    result = await coro
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/client/utilities.py:69: in with_injected_client
    client, inferred = get_or_create_client(client)
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/client/utilities.py:52: in get_or_create_client
    return get_httpx_client(), False
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/client/orchestration.py:176: in get_client
    from prefect.server.api.server import create_app
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/server/__init__.py:1: in <module>
    from . import models, orchestration, schemas, services
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/server/services/__init__.py:7: in <module>
    import prefect.server.services.task_scheduling
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/server/services/task_scheduling.py:12: in <module>
    from prefect.server.api.task_runs import TaskQueue
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/server/api/__init__.py:1: in <module>
    from . import (
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/server/api/deployments.py:30: in <module>
    from prefect.utilities.schema_tools.validation import (
../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/utilities/schema_tools/__init__.py:2: in <module>
    from .validation import CircularSchemaRefError, ValidationError, validate
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    from collections import defaultdict, deque
    from copy import deepcopy
    from typing import Dict, List
    
    import jsonschema
    from jsonschema.exceptions import ValidationError as JSONSchemaValidationError
>   from jsonschema.validators import Draft202012Validator, create
E   ImportError: cannot import name 'Draft202012Validator' from 'jsonschema.validators' (/home/aradox/.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/jsonschema/validators.py)

../../../.pyenv/versions/3.10.9/envs/prefect/lib/python3.10/site-packages/prefect/utilities/schema_tools/validation.py:7: ImportError
```


### Versions

```Text
Version:             2.16.9
API version:         0.8.4
Python version:      3.10.9
Git commit:          083def52
Built:               Thu, Apr 4, 2024 3:11 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         cloud
```


### Additional context

_No response_

--- a/requirements-client.txt
+++ b/requirements-client.txt
@@ -12,7 +12,7 @@ httpx[http2] >= 0.23, != 0.23.2
 importlib_metadata >= 4.4; python_version < '3.10'
 importlib-resources >= 6.1.3, < 6.2.0
 jsonpatch >= 1.32, < 2.0
-jsonschema >= 3.2.0, < 5.0.0
+jsonschema >= 4.0.0, < 5.0.0
 orjson >= 3.7, < 4.0
 packaging >= 21.3, < 24.3
 pathspec >= 0.8.0

